Added wildcard support to groovyw <item> get

BenjaminAmos opened the pull request:
<!-- Thanks for submitting a pull request for Destination Sol! :-)
Please fill in some brief details below about the PR.
If it contains source code please make sure to do everything in the pre pull request checklist first.
If you add unit tests we'll love you forever! -->

# Description
This PR adds wildcard support to the `groovyw <item> get <object>` e.g. `groovyw <item> get *`.

# Testing
Testing this PR could take some considerable time and effort to check all of the use-cases are correct, so these changes should only apply if wildcards (`*` or `?`) are detected in the input argument (see [Line 61 of util.groovy](https://github.com/BenjaminAmos/DestinationSol/blob/6a9376b3216067f14163f930dc873fc713865f9c/config/groovy/util.groovy#L61)). As such, I will list the series of tests that I have tried so far:
## Single argument tests:
- [X] Command: `groovyw module get *`
    - Expected result: All the available modules under the @DestinationSol organisation should be obtained.
- [X] Command: `groovyw module get d??p*`
    - Expected result: This should obtain the [deepspace](https://github.com/DestinationSol/deepspace) module.
- [X] Command: `groovyw module get s??????`
    - Expected result: This should obtain the [salvage](https://github.com/DestinationSol/salvage) module.
- [X] Command: `groovyw module get f?rm?c`
    - Expected result: This should obtain the [formic](https://github.com/DestinationSol/formic) module.
- [X] Command: `groovyw module get t*`
    - Expected result: This should obtain the [tribe](https://github.com/DestinationSol/tribe) module.
- [X] Command: `groovyw module get *de*`
    - Expected result: This should obtain both the [deepspace](https://github.com/DestinationSol/deepspace) and the [federal](https://github.com/DestinationSol/federal) modules.
## Multiple argument tests:
- [X] Command: `groovyw module get *de* fo*`
    - Expected result: This should obtain both the [deepspace](https://github.com/DestinationSol/deepspace), [federal](https://github.com/DestinationSol/federal) modules, as well as the [deepspace](https://github.com/DestinationSol/deepspace) and the [formic](https://github.com/DestinationSol/formic) module.
- [X] Command: `groovyw module get f* d???*`
    - Expected result: This should obtain the [deepspace](https://github.com/DestinationSol/deepspace), [draconic](https://github.com/DestinationSol/draconic), [federal](https://github.com/DestinationSol/federal) and [formic](https://github.com/DestinationSol/formic) modules.
## Testing regex characters:
- [X] Command `groovyw module get .`
    - Expected result: This should attempt to obtain the "." module, which does not exist.
- [X] Command `groovyw module get (`
    - Expected result: This should attempt to obtain the "(" module, which does not exist.
- [X] Command `groovyw module get )`
    - Expected result: This should attempt to obtain the ")" module, which does not exist.
- [X] Command `groovyw module get []`
    - Expected result: This should attempt to obtain the "[]" module, which does not exist.
- [X] Command `groovyw module get *`
    - Expected result: This character is specially handled, so all the available modules under the @DestinationSol organisation should be obtained.
## Testing multiple wildcards in succession:
- [X] Command `groovyw module get ??????`
    - Expected result: This should attempt to obtain the [formic](https://github.com/DestinationSol/formic), [slimey](https://github.com/DestinationSol/slimey) and [vulcan](https://github.com/DestinationSol/vulcan) modules.
- [X] Command `groovyw module get ******`
    - Expected result: All the available modules under the @DestinationSol organisation should be obtained.

# Outstanding Work
- [X] Such a large change needs significant testing. This change needs to be tested extensively to be proven useable, although I think that it should work most of the time. I still need to finish doing this, as I have not fully tested it yet.
    - Further tests to perform:
        - [X] Test that the regex characters (``[``, ``]``, ``(``, ``)``, ``*``, ``.``, ``\``) are not injected into the code as regexes
        - [X] Test if multiple wildcard selectors can be used to obtain completely different modules, e.g. using the command `groovyw module get *de* fo*`, which should obtain both the [deepspace](https://github.com/DestinationSol/deepspace), [federal](https://github.com/DestinationSol/federal) modules, as well as the [deepspace](https://github.com/DestinationSol/deepspace) and the [formic](https://github.com/DestinationSol/formic) module.

# Notes
This should not break any existing functionality, as the new wildcard handling code is only used when the `*` or `?` symbols are found.

The wildcard strings will have to be escaped using single quotes (e.g. `./groovyw module get '*'` rather than `./groovyw module get *`) on POSIX platforms (Linux and MacOS), as otherwise the terminal appears to automatically expand them (replacing the arguments with the names all the files in the current directory matching that descriptor) before passing them to the program. I have modified the Windows script to automatically escape the arguments, although I am not sure if it is possible with POSIX platforms to do so.

<!--
### Pre Pull Request Checklist:
 When scanning the code with SonarLint, just don't introduce any new issues, and maybe even fix a few existing ones
- [ ] Code has been scanned with [SonarLint](http://www.sonarlint.org/intellij/)
- [ ] Methods have been annotated with @ Nullable and @ Nonnull ([Read more](https://github.com/MovingBlocks/DestinationSol/wiki/@Nullable-&-@Nonnull-Quickstart))
- [ ] There are no errors present in the project
- [ ] Code has been formatted and indented
- [ ] Methods have appropriate Javadoc ([How to write Javadoc](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html))
-->

GooeyHub commented:
Can one of the admins please verify this patch?

Steampunkery commented:
Hey, looks nice frome here--haven't tested, just glanced at the code. Can your rebase onto develop for me? There were some groovy-related module fetching PRs merged between when you started this branch and now. I tried locally, there shouldn't be any rebase conflicts.

BenjaminAmos commented:
The commits should now be rebased onto develop.

arpitkamboj commented:
So I've tested this extensively and all the features work as intended but there is one small problem:

The `?` wildcard only functions when it is enclosed between two non-wildcard characters

For example:
`groovyw module get ?` does the same thing as `groovyw module get *`
`groovyw module get s??????` downloads the salvage module but so does `groovyw module get s?`
`*n` and `?n` both correspond to modules ending with `n` 
`ca???on` is working as intended and retrieves `caution` and `c?n` does not retrieve anything which is good.

Please fix this issue. Everything else is working as intended. Nice Work!

BenjaminAmos commented:
The issue should be fixed now. Thank you for testing this. I believe, after further testing myself, that this change, the wildcards should work now in most user-supplied cases.

arpitkamboj commented:
Hey!
So the issue with the `?` wildcard is now fixed but there is another weird thing.
I don't know if this is really important though but still:
`groovyw module get ??????` tries to get the modules with the file names in the local directory having 6 characters (eg config, engine, gradle) instead of getting modules having names with 6 characters

Please try if you can fix this issue
Thanks for the great work!

Adrijaned commented:
Hey, @arpitkamboj , I'm afraid the issue you are describing arises from `sh` itself, as it replaces the ?s on terminal with the matching local files itself, and does not even pass the ?s to the script. For fixing this, you must simply surround the ?s with `"`

BenjaminAmos commented:
This issue should be fixed on Windows, for the arguments are preserved as-is and I just needed to escape them before passing them as arguments to the java executable. I was doing the same for the "*" character anyway. I can't do anything about the sh script though, as the parameter expansion seems to happen before the arguments are passed to the script.

Thanks for the advice @Adrijaned. I have added a note about this to my original message and added test-cases there so that I am not caught out by that issue again.

vampcat commented:
@GooeyHub : add to whitelist

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Cervator commented:
Bump - I've reviewed this on the Terasology side and wanted to refer to the comments from here since they likely relate on both sides. Somewhat concerned about the OS complexities introduced by the wildcarding on command line. https://github.com/MovingBlocks/Terasology/pull/3531

Otherwise agreed this is very good work, especially the thorough testing, so thank you so much for that :-)

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Cervator commented:
Update: I have added to this for Terasology and tested it out fairly thoroughly. We should be able to merge this here then override with the related files from there, double check / lightly test, then merge.

Was aiming to do that now but I hit #414 on my main test system while out of town. Should be easy to do later or for somebody else. See https://github.com/MovingBlocks/Terasology/pull/3531 for details along with #3679 which I also merged and https://github.com/MovingBlocks/Terasology/commit/dbc32f0da97e46c447ae2293a00e5babaf92c85b which I added on top.
